pkg/controller/cloud/node_gc_controller.go

```diff
@@ -0,0 +1,214 @@
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/record"
+	cloudprovider "k8s.io/cloud-provider"
+	nodeutilv1 "k8s.io/kubernetes/pkg/api/v1/node"
+	"k8s.io/kubernetes/pkg/controller"
+	schedulerapi "k8s.io/kubernetes/pkg/scheduler/api"
+)
+
+var ShutdownTaint = &v1.Taint{
+	Key:    schedulerapi.TaintNodeShutdown,
+	Effect: v1.TaintEffectNoSchedule,
+}
+
+// CloudNodeGCController is responsible for garbage collecting kubernetes
+// nodes that have been deleted/shutdown on the cloud provider
+type CloudNodeGCController struct {
+	kubeClient clientset.Interface
+	recorder   record.EventRecorder
+
+	cloud cloudprovider.Interface
+
+	// Value controlling NodeController monitoring period, i.e. how often does NodeController
+	// check node status posted from kubelet. This value should be lower than nodeMonitorGracePeriod
+	// set in controller-manager
+	nodeMonitorPeriod time.Duration
+}
+
+func NewCloudNodeGarbageCollectorController(
+	kubeClient clientset.Interface,
+	cloud cloudprovider.Interface,
+	nodeMonitorPeriod time.Duration) (*CloudNodeGCController, error) {
+
+	eventBroadcaster := record.NewBroadcaster()
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "cloud-node-gc-controller"})
+	eventBroadcaster.StartLogging(glog.Infof)
+
+	glog.V(0).Infof("Sending events to api server.")
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+
+	if kubeClient == nil {
+		glog.Fatal("kube client is nil while starting cloud node deletion controller")
+	}
+
+	if cloud == nil {
+		return nil, errors.New("no cloud provider provided")
+	}
+
+	if _, ok := cloud.Instances(); !ok {
+		return nil, errors.New("cloud provider does not support instances")
+	}
+
+	c := &CloudNodeGCController{
+		kubeClient:        kubeClient,
+		recorder:          recorder,
+		cloud:             cloud,
+		nodeMonitorPeriod: nodeMonitorPeriod,
+	}
+
+	return c, nil
+}
+
+func (c *CloudNodeGCController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	go wait.Until(c.GarbageCollect, c.nodeMonitorPeriod, stopCh)
+}
+
+func (c *CloudNodeGCController) GarbageCollect() {
+	instances, ok := c.cloud.Instances()
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("failed to get instances from cloud provider"))
+		return
+	}
+
+	nodes, err := c.kubeClient.CoreV1().Nodes().List(metav1.ListOptions{ResourceVersion: "0"})
+	if err != nil {
+		glog.Errorf("Error monitoring node status: %v", err)
+		return
+	}
+
+	for i := range nodes.Items {
+		var currentReadyCondition *v1.NodeCondition
+		node := &nodes.Items[i]
+		// Try to get the current node status
+		// If node status is empty, then kubelet has not posted ready status yet. In this case, process next node
```

Comment:
What if the node was deleted or shut down in the cloud provider before the Kubelet could post a ready status?

It might be more robust to assume `v1.ConditionUnknown` if there is no ready condition on the node yet, and then call directly into the cloud to see if a VM exists/is shut down. Since in the unknown case we'd then always check the cloud provider directly, and only take action when the cloud reports deletion or shut-down, we can probably also avoid the retries/additional Get requests. 

And then if we're also listing from the informer cache, that will help to reduce any additional load on the API server that splitting out these control loops would create.

If we then went a little further and moved the body of this for loop into a helper function, the initial status checks might collapse to something like this, at which point we can un-nest a lot of the subsequent logic:
```
// Default status to v1.ConditionUnknown (in fact, this is the same as the default "fake" status from tryUpdateNodeHealth in the kube node lifecycle controller)
status := v1.ConditionUnknown
if _, c := v1node.GetNodeCondition(&node.Status, v1.NodeReady); c != nil {
  status = c.Status
}

// Skip healthy nodes.
if status == v1.ConditionTrue {
  return
}
```